### PR TITLE
NC | Online Upgrade | Health CLI update config directory and upgrade checks

### PR DESCRIPTION
**Pull request by romayalon**
### Explain the changes
1. Added a check in the Health CLI that shows the config directory status if exists.
2. Added a check in the Health CLI that asserts if the upgrade of the config directory has issues, this situation can be discovered by phase of the config directory is locked and there is no ongoing upgrade.
3. Added unit tests (and did some refactoring).
4. Fixed a bug that sets the process name to be `Upgrade`, it's showing in the logs of the endpoint, CLI etc.

### Issues: Fixed #xxx / Gap #xxx
1. Another idea for a health check - check per host if config directory is locked.

### Testing Instructions:
1.  `sudo node ./node_modules/mocha/bin/mocha src/test/unit_tests/test_nc_health.js`

- [x] Doc added/updated
- [x] Tests added

### CODE REVIEW

File: src/manage_nsfs/health.js [hunk 1]
```diff
@@ -421,6 +425,38 @@ class NSFSHealth {
...
+     */
+    _get_config_dir_status(system_data) {
+        if (!system_data) return { error: 'system data is missing' };
+        const config_dir_data = system_data.config_directory;
+        if (!config_dir_data) return { error: 'config directory data is missing' };
+        return {
+            phase: config_dir_data.phase,
+            config_dir_version: config_dir_data.config_dir_version,
+            upgrade_package_version: config_dir_data.upgrade_package_version,
+            in_progress_upgrade: config_dir_data.in_progress_upgrade
```
Comment (naveenpaul1) [inline src/manage_nsfs/health.js:442]: Do we need to add error type for config and upgrad status?

File: src/manage_nsfs/health.js [hunk 2]
```diff
@@ -136,6 +138,8 @@ class NSFSHealth {
                     endpoint_state,
                     error_type: health_errors_tyes.TEMPORARY,
                 },
+                config_directory_status: this._get_config_dir_status(system_data),
+                upgrade_status: this._get_upgrade_status(system_data),
```
Comment (naveenpaul1) [inline src/manage_nsfs/health.js:142]: update health doc for newly added properties.

File: src/manage_nsfs/health.js [hunk 3]
```diff
@@ -421,6 +425,38 @@ class NSFSHealth {
...
+            phase: config_dir_data.phase,
+            config_dir_version: config_dir_data.config_dir_version,
+            upgrade_package_version: config_dir_data.upgrade_package_version,
+            in_progress_upgrade: config_dir_data.in_progress_upgrade
+        };
+    }
+
+    /**
+     * get_upgrade_status returns the status of an ongoing upgrade, if valid it returns an object with upgrade details
+     * if upgrade is not ongoing but config dir is locked in one of the hosts, the error details will return
```
Comment (shirady) [inline src/manage_nsfs/health.js:448]: Where do we return the "an object with upgrade details" I don't see it?

File: src/test/unit_tests/test_nc_health.js [hunk 4]
```diff
@@ -45,6 +56,10 @@ mocha.describe('nsfs nc health', function() {
     });
 
     mocha.describe('nsfs nc health cli validations', function() {
+        mocha.afterEach(() => {
```
Comment (shirady) [inline src/test/unit_tests/test_nc_health.js:59]: It was at the beginning of each it, right? why not use `beforeEach`?

File: src/test/unit_tests/test_nc_health.js [hunk 5]
```diff
@@ -158,13 +173,20 @@ mocha.describe('nsfs nc health', function() {
 
         mocha.afterEach(async () => {
             await fs_utils.file_delete(config_fs.config_json_path);
+            restore_health_if_needed(Health);
         });
 
-        mocha.it('Health all condition is success', async function() {
-            const get_service_state = sinon.stub(Health, "get_service_state");
+        mocha.it('Health all condition is success1', async function() {
```
Comment (shirady) [inline src/test/unit_tests/test_nc_health.js:179]: Was this "1" intentional? 
```suggestion
        mocha.it('Health all condition is success', async function() {
```

File: src/test/unit_tests/test_nc_health.js [hunk 6]
```diff
@@ -601,19 +586,82 @@ mocha.describe('nsfs nc health', function() {
...
-            const get_endpoint_response = sinon.stub(Health, "get_endpoint_response");
+            const get_endpoint_response = sinon.stub(Health, 'get_endpoint_response');
             get_endpoint_response.onFirstCall().returns(Promise.resolve({response: {response_code: 'RUNNING', total_fork_count: 0}}));
             const health_status = await Health.nc_nsfs_health();
             assert.strictEqual(health_status.checks.accounts_status.valid_accounts.length, 1);
             assert.strictEqual(health_status.checks.accounts_status.invalid_accounts.length, 1);
             await exec_manage_cli(TYPES.ACCOUNT, ACTIONS.DELETE, { config_root, name: account_invalid.name });
         });
+
+        mocha.it('Health all condition - invalid config directory upgrade', async function() {
```
Comment (shirady) [inline src/test/unit_tests/test_nc_health.js:602]: Would you add a comment - what makes it invalid?

Comment (romayalon) [reply src/test/unit_tests/test_nc_health.js]: invalid means that the upgrade failed, I rephrased the test's headline

File: src/manage_nsfs/health.js (hunk 3 above)
Comment (romayalon) [reply src/manage_nsfs/health.js]: hmm maybe it was missing when you reviewed it, it's there now

File: src/manage_nsfs/health.js (hunk 1 above)
Comment (romayalon) [reply src/manage_nsfs/health.js]: sure

File: src/test/unit_tests/test_nc_health.js [hunk 7]
```diff
@@ -601,19 +591,107 @@ mocha.describe('nsfs nc health', function() {
...
+                'upgrade_package_version': pkg.version,
+                'phase': CONFIG_DIR_LOCKED,
+                upgrade_history: {
+                    successful_upgrades: [],
+                    last_failure: { error: 'mock error'}
+                }
+            };
+            await Health.config_fs.create_system_config_file(JSON.stringify(valid_system_json));
+            const health_status = await Health.nc_nsfs_health();
+            await fs_utils.file_delete(Health.config_fs.system_json_path);
```
Comment (shirady) [inline src/test/unit_tests/test_nc_health.js:619]: Why do we delete this? Do we do it in every test? if so we can add it to `afterEach.

File: src/test/unit_tests/test_nc_health.js [hunk 8]
```diff
@@ -601,19 +591,107 @@ mocha.describe('nsfs nc health', function() {
...
+
+        mocha.it('Health all condition - valid ongoing config directory upgrade', async function() {
+            valid_system_json.config_directory = {
+                config_dir_version: config_fs.config_dir_version,
+                upgrade_package_version: pkg.version,
+                phase: CONFIG_DIR_LOCKED,
+                upgrade_history: {
+                    successful_upgrades: [],
+                },
+                in_progress_upgrade: { from_version: 'x', to_version: 'y'}
```
Comment (shirady) [inline src/test/unit_tests/test_nc_health.js:631]: Maybe add a real version details?
```suggestion
                in_progress_upgrade: { from_version: '5.17.8', to_version: '5.18.0'}
```

**Review by shirady — APPROVED** [review]
LGTM
I added some comments and question which are minor and can be changed in future PRs.

File: src/test/unit_tests/test_nc_health.js (hunk 7 above)
Comment (romayalon) [reply src/test/unit_tests/test_nc_health.js:619]: No, we don't do it on every test, on most tests we use mock system json, and on specific tests in which I want to test specifically the health of the upgrade/config dir based on the system json I actually create and delete this file.

File: src/manage_nsfs/health.js [hunk 9]
```diff
@@ -173,13 +183,23 @@ class NSFSHealth {
...
+    get_error_code({ service_status, pid, endpoint_response_code, config_directory_status }) {
+        if (service_status !== 'active' || pid === '0') {
             return health_errors.NOOBAA_SERVICE_FAILED;
         } else if (endpoint_response_code === 'NOT_RUNNING') {
             return health_errors.NOOBAA_ENDPOINT_FAILED;
         } else if (endpoint_response_code === 'MISSING_FORKS') {
             return health_errors.NOOBAA_ENDPOINT_FORK_MISSING;
+        } else if (config_directory_status.error) {
+            return health_errors.CONFIG_DIR_ERROR;
         }
```
[the remark this answers is not in the document]
Comment (romayalon) [reply src/manage_nsfs/health.js:203]: nope, the default is no error = undefined

File: src/manage_nsfs/health.js [hunk 10]
```diff
@@ -161,7 +171,7 @@ class NSFSHealth {
                 delay_ms: config.NC_HEALTH_ENDPOINT_RETRY_DELAY,
                 func: async () => {
                     endpoint_state = await this.get_endpoint_fork_response();
-                    if (endpoint_state.response.response_code === fork_response_code.NOT_RUNNING.response_code) {
+                    if (endpoint_state.response?.response_code === fork_response_code.NOT_RUNNING.response_code) {
```
[the remark this answers is not in the document]
Comment (romayalon) [reply src/manage_nsfs/health.js:174]: I added it to https://github.com/noobaa/noobaa-core/issues/8318
please add there if you have more ideas that are unrelated to these changes

File: src/manage_nsfs/health.js [hunk 11]
```diff
@@ -421,6 +441,62 @@ class NSFSHealth {
...
+     * endpoint_response_code: string, 
+     * config_directory_status: Object }} health_check_params
+     * @returns {'OK' | 'NOTOK'}
+     */
+    _calc_health_status({ service_status, pid, endpoint_response_code, config_directory_status }) {
+        const is_unhealthy = service_status !== 'active' ||
+            pid === '0' ||
+            endpoint_response_code !== 'RUNNING' ||
+            config_directory_status.error;
+        return is_unhealthy ? 'NOTOK' : 'OK';
```
[the remark this answers is not in the document]
Comment (romayalon) [reply src/manage_nsfs/health.js:498]: I added it to https://github.com/noobaa/noobaa-core/issues/8318
please add there if you have more ideas that are unrelated to these changes

File: src/manage_nsfs/health.js [hunk 12]
```diff
@@ -117,12 +122,16 @@ class NSFSHealth {
...
+        const system_data = await this.config_fs.get_system_config_file({ silent_if_missing: true });
+        const config_directory_status = this._get_config_dir_status(system_data);
+
         let bucket_details;
         let account_details;
-        const response_code = endpoint_state ? endpoint_state.response.response_code : 'NOT_RUNNING';
-        const service_health = service_status !== 'active' || pid === '0' || response_code !== 'RUNNING' ? 'NOTOK' : 'OK';
-
-        const error_code = await this.get_error_code(service_status, pid, response_code);
+        const endpoint_response_code = endpoint_state ? endpoint_state.response?.response_code : 'NOT_RUNNING';
```
[the remark this answers is not in the document]
Comment (romayalon) [reply src/manage_nsfs/health.js:131]: it was actually a fix for a bug I saw when running Health CLI while server is down, 
I thought about your question and I think that the previous decision to make the default NOT_RUNNING is wrong and it should be UNKNOWN, so I changed it appropiatly.

File: src/manage_nsfs/health.js [hunk 13]
```diff
@@ -421,6 +441,62 @@ class NSFSHealth {
...
+
+    /**
+     * _get_config_dir_status returns the config directory phase, version and equivalent package_version
+     * @param {Object} system_data 
+     * @returns {Object}
+     */
+    _get_config_dir_status(system_data) {
+        if (!system_data) return { error: 'system data is missing' };
+        const config_dir_data = system_data.config_directory;
+        if (!config_dir_data) return { error: 'config directory data is missing' };
```
[the remark this answers is not in the document]
Comment (romayalon) [reply src/manage_nsfs/health.js:453]: not sure if you saw it but I did some changes and now the config dir upgrade status is a part of the config dir status so the second check is redundant completely

File: src/test/unit_tests/test_nc_health.js [hunk 14]
```diff
@@ -18,11 +19,30 @@ const { get_process_fs_context } = require('../../util/native_fs_utils');
 const { ManageCLIError } = require('../../manage_nsfs/manage_nsfs_cli_errors');
 const { TYPES, DIAGNOSE_ACTIONS, ACTIONS } = require('../../manage_nsfs/manage_nsfs_constants');
 const { TMP_PATH, create_fs_user_by_platform, delete_fs_user_by_platform, exec_manage_cli } = require('../system_tests/test_utils');
+const { CONFIG_DIR_UNLOCKED, CONFIG_DIR_LOCKED } = require('../../upgrade/nc_upgrade_manager');
```
[the remark this answers is not in the document]
Comment (romayalon) [reply src/test/unit_tests/test_nc_health.js:22]: unrelated to the Health script, I added it to here https://github.com/noobaa/noobaa-core/issues/8586

**Pull request merged**